# Data Encryption Standard {DES}
- Created 1970 by IBM

- Symmetric Encryption 

- (Block Ciphering Algorithm) => 64 Bit Standard

- Complex => {Subistitution , Transposition}

- starting by chopping the original message to 64-bit Blocks

- The Key is 8 bits added to each block for check {56 + 8} = 64

- 16 iterations are involved.

## There are 3 Types of DES
- DES 

    - In 1998, a US$ 200,000 “DES cracker” machine was invented to crack any DES in 4 days. 

    - Later, DES is deciphered in few hours. 

- Double DES or 2DES

    - Double Key DES was invented later followed by Triple DES.  

    - Double DES uses the following method: $𝐸(𝑘2,𝐸(𝑘1,𝑚))$

    - In theory, double DES multiply the difficulty of breaking the code

    - Merkle and Hellman double DES proved that using double `56bits key` results in similar result as if we were using single `57bits key`. 
    - Therefore, DES `adds no security`. 
    
    - Breaking double DES requires 2000 years

- Triple DES or 3DES

    - The cipher texted is produced using 3DES as follows: $𝐸(𝑘3,𝐸(𝑘2,𝐸(𝑘1,𝑚)))$

    - The key produced is roughly equivalent to `112-bit key`
    - Two-Key 3DES is invented using the following formula: $𝐸(𝑘1,𝐷(𝑘2,𝐸(𝑘1,𝑚)))$
    
    - The key produced is roughly equivalent to `80-bit` key.
    - Breaking a 3DES is significantly longer than 2DES.

    - In 1995, Advanced Encryption Standered (AES) was introduced. 

    - In 2001, AES became the standard encryption for all federal government. 


    



![](PICS/des.gif)

## Now Let's Work on DES using PyPI to install packages
```
pip install des
```


In [1]:
from des import DesKey

- The key should be of length 8, 16 or 24. The algorithm will be automatically chosen for you. 

- In simple Way if you need just the DES write only a key with 8 bytes or less , But if less it will be repeated
## Example

In [2]:
key0 = DesKey(b"some key")                  # for DES 8 byte ==> 64 bit
key1 = DesKey(b"a key for TRIPLE")          # for 3DES, same as "a key for TRIPLEa key fo"
key2 = DesKey(b"a 24-byte key for TRIPLE")  # for 3DES
key3 = DesKey(b"1234567812345678REAL_KEY")  # for DES, same as "REAL_KEY"

- To Check the chosen algorithm 
    - `is_single()`
    - `is_triple()`
    

In [4]:
print(key0.is_single())
print(key0.is_triple())
print(key2.is_triple())

True
False
True


To encrypt use `encrypt()` and to `decrypt()` use `decrypt()`

In [15]:
%%time
CipherText1 = key0.encrypt(b"any long message that have been occubied to you , have a good luck" , padding=True)
print(CipherText1)

b"\x14\xfa\xc2 '\x00{\xa9\xdc;\x9dq\xcbr\x87Q"
Wall time: 4.05 ms


In [17]:
%%time
PlainText1 = key0.decrypt(CipherText1);
PlainText1

Wall time: 4 ms


b'any long message that have been occubied to you , have a good luck\x06\x06\x06\x06\x06\x06'

- This Message Use 4 ms to encrypt or decrypt
- Let's check the 3DES

In [28]:
%%time
CipherText2 = key2.encrypt(b"any long message that have been occubied to you , have a good luck" , padding=True)
CipherText2

Wall time: 13 ms


b'QOZ\x00\xc1\x17\x92\x07\xbf\xb3~z\xc5\xa9\x90\xff\x07\x81\xea\x9er\x1d\r\xe2P<J\xd7\xc5\x08Ox\xbc )ZY\xc30\x7f\x9e\x9fn\x98\xaa\xc4_\x11%\xb7i\xafJ)\xaeh\xb9\xdeY\xa0\x99\xc8\x12_\x06\t8\xea\x17\x80\xcd\xc3'

In [29]:
%%time
PlainText2 = key2.decrypt(CipherText2);
PlainText2

Wall time: 14 ms


b'any long message that have been occubied to you , have a good luck\x06\x06\x06\x06\x06\x06'

In [30]:
14/4


3.5

So the `DES` is faster than the `3DES` 3.5 times 